File: Machine_Learning_A_Z/Part 7 - Natural Language Processing/1 - NLP.R
# NLP!!!!!!!

# Importing the dataset
dataset_original <- read.delim("Restaurant_Reviews.tsv", quote = "", stringsAsFactors = FALSE)



# CLEANING THE DATASET
#install.packages("tm")
library(tm)
#install.packages("SnowballC")                    #For stopwords
library(SnowballC)
# as.character(corpus[[1]])
corpus <- VCorpus(VectorSource(dataset$Review))
corpus <- tm_map(corpus, content_transformer(tolower))
# as.character(corpus[[841]])
corpus <- tm_map(corpus, removeNumbers)
# as.character(corpus[[841]])
corpus <- tm_map(corpus, removePunctuation)
corpus <- tm_map(corpus, removeWords, stopwords())
# as.character(corpus[[1]])
corpus <- tm_map(corpus, stemDocument)
# as.character(corpus[[1]])
corpus <- tm_map(corpus, stripWhitespace)
# as.character(corpus[[841]])



# CREATING THE BAG OF WORDS MODEL
# Sparce matrix / Sparcity / Tokenization
dtm <- DocumentTermMatrix(corpus)
dtm <- removeSparseTerms(dtm, sparse = 0.999)           #99.9% of the most frequent (99.9% of the most 1s)
dtm
dataset <- as.data.frame(as.matrix(dtm))
dataset$Liked <- dataset_original$Liked



# FITTING USING RANDOM FOREST
# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest Classification to the Training set
library(randomForest)
classifier = randomForest(training_set[-692], training_set$Liked,
                          ntree = 10)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
cm

82+77/200

File: Machine_Learning_A_Z/Part 7 - Natural Language Processing/1 - NLP.py
# NLP!!!!!!!

# Importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt



# Importing the dataset
dataset = pd.read_csv("Restaurant_Reviews.tsv", delimiter = '\t', quoting = 3)



# CLEANING THE TEXT
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

# Remove all NON-LETTERS and replace by a ' ' (Space)
review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][0])
# Convert to lower case and split into a LIST
review = review.lower()
review = review.split()
# review = [word for word in review if word not in set(stopwords.words('english'))]
# We use a "SET" because the format of "LIST" that it is in would be slower to read from

# STEMMING
ps = PorterStemmer()
review = [ps.stem(word) for word in review if word not in set(stopwords.words('english'))]

# Join the values
review = ' '.join(review)



# LOOPING FOR ALL VALUES IN THE DATASET
corpus = []
# A corpus in ML is a term given to a collection of texts like reviews or longer like,
    # web scrapped, articles, etc.
for i in range(0,1000):
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
    review = review.lower()
    review = review.split()
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if word not in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)
    
    
   
# CREATING THE BAG OF WORDS MODEL
    #Sparce matrix / Sparcity / Tokenization
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
# We can perform most of the cleaning operationg using the various parameters of the
    # CountVectorizer class
# MAX_FEATURES is the (in this case) 1500 most common features 
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, 1].values




# USING NAIVE BAYES TO FIT THE MODEL
# We mostly use Naive Bayes, Decision Tree, or Random Forests for NLP 
# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

# Fitting Naive Bayes to the Training set
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
cm

accuracy = (cm[0,0] + cm[1,1]) / [len(dataset) * 0.2]           # TEST_SIZE = 0.2
accuracy









